File: M3Tools/Controls/People/Users/ConfirmPasswordField.vb
Imports System.Windows.Forms

Public Class ConfirmPasswordField
    Public Event ConfirmLostFocus As EventHandler
    Public Event ConfirmTextChanged As EventHandler
    Public Shadows Property Text As String
        Get
            Return txt_ConfirmPassword.Text
        End Get
        Set(value As String)
            txt_ConfirmPassword.Text = value
        End Set
    End Property

    ''' <summary>
    ''' Gets the length of text in text control.
    ''' </summary>
    ''' <returns>The number of characters contained in the text of the control.</returns>
    Public ReadOnly Property TextLength As Integer
        Get
            Return txt_ConfirmPassword.TextLength
        End Get
    End Property

    Public Overrides ReadOnly Property Focused As Boolean
        Get
            Return txt_ConfirmPassword.Focused
        End Get
    End Property

    Public ReadOnly Property ConfirmField As TextBox
        Get
            Return txt_ConfirmPassword
        End Get
    End Property


	Private Sub Btn_Show_Click(sender As Object, e As EventArgs) Handles btn_Show.Click
		'Switch between password icon image for the button
		btn_Show.Image = If(txt_ConfirmPassword.UseSystemPasswordChar, My.Resources.HidePasswordIcon, My.Resources.ShowPasswordIcon)

		'Invert the use system password char property
		txt_ConfirmPassword.UseSystemPasswordChar = Not txt_ConfirmPassword.UseSystemPasswordChar
	End Sub

	''' <summary>
	''' Clears all text from the text box control.
	''' </summary>
	Public Sub Clear()
		txt_ConfirmPassword.Clear()
	End Sub

	''' <summary>
	''' Selects a range of text in the text box.
	''' </summary>
	''' <param name="start"></param>
	''' <param name="length"></param>
	''' <exception cref="ArgumentOutOfRangeException"></exception>
	Public Shadows Sub [Select](start As Integer, length As Integer)
		txt_ConfirmPassword.Select(start, length)
	End Sub

	''' <summary>
	''' Selects all text in the text box.
	''' </summary>
	Public Shadows Sub SelectAll()
		txt_ConfirmPassword.SelectAll()
		'txtPassword.Select(0, TextLength)
	End Sub

	Public Overloads Function Focus() As Boolean
		Return txt_ConfirmPassword.Focus()
	End Function

	Private Sub Txt_ConfirmPassword_LostFocus(sender As Object, e As EventArgs) Handles txt_ConfirmPassword.LostFocus
		RaiseEvent ConfirmLostFocus(Me, e)
	End Sub

	Private Sub Txt_ConfirmPassword_TextChanged(sender As Object, e As EventArgs) Handles txt_ConfirmPassword.TextChanged
		RaiseEvent ConfirmTextChanged(Me, e)
	End Sub
End Class

File: M3Tools/Controls/People/Users/ConfirmPasswordField.cs
using System;

namespace SPPBC.M3Tools
{

	public partial class ConfirmPasswordField
	{
		/// <summary>
		/// The event that occurs when the confirm field has lost focus
		/// </summary>
		public event EventHandler ConfirmLostFocus;

		/// <summary>
		/// The event that occurs when the confirm field has been changed
		/// </summary>
		public event EventHandler ConfirmTextChanged;

		/// <summary>
		/// The value of the confirm field
		/// </summary>
		public new string Text
		{
			get => txt_ConfirmPassword.Text;
			set => txt_ConfirmPassword.Text = value;
		}

		/// <summary>
		/// Gets the length of text in text control.
		/// </summary>
		/// <returns>The number of characters contained in the text of the control.</returns>
		public int TextLength => txt_ConfirmPassword.TextLength;

		/// <summary>
		/// Whether the control is focused
		/// </summary>
		public override bool Focused => txt_ConfirmPassword.Focused;

		/// <summary>
		/// 
		/// </summary>
		public ConfirmPasswordField()
		{
			InitializeComponent();
		}

		private void Btn_Show_Click(object sender, EventArgs e)
		{
			// Switch between password icon image for the button
			btn_Show.Image = txt_ConfirmPassword.UseSystemPasswordChar ? Properties.Resources.HidePasswordIcon : Properties.Resources.ShowPasswordIcon;

			// Invert the use system password char property
			txt_ConfirmPassword.UseSystemPasswordChar = !txt_ConfirmPassword.UseSystemPasswordChar;
		}

		/// <summary>
		/// 	''' Clears all text from the text box control.
		/// 	''' </summary>
		public void Clear()
		{
			txt_ConfirmPassword.Clear();
		}

		/// <summary>
		/// 	''' Selects a range of text in the text box.
		/// 	''' </summary>
		/// 	''' <param name="start"></param>
		/// 	''' <param name="length"></param>
		/// 	''' <exception cref="ArgumentOutOfRangeException"></exception>
		public void Select(int start, int length)
		{
			txt_ConfirmPassword.Select(start, length);
		}

		/// <summary>
		/// 	''' Selects all text in the text box.
		/// 	''' </summary>
		public void SelectAll()
		{
			txt_ConfirmPassword.SelectAll();
			// txtPassword.Select(0, TextLength)
		}

		/// <summary>
		/// Focus on the confirm field
		/// </summary>
		/// <returns></returns>
		public new bool Focus()
		{
			return txt_ConfirmPassword.Focus();
		}

		private void Txt_ConfirmPassword_LostFocus(object sender, EventArgs e)
		{
			ConfirmLostFocus?.Invoke(this, e);
		}

		private void Txt_ConfirmPassword_TextChanged(object sender, EventArgs e)
		{
			ConfirmTextChanged?.Invoke(this, e);
		}
	}
}
